timemachine/potentials/generic.py

```diff
@@ -0,0 +1,236 @@
+from dataclasses import dataclass
+from typing import Any, Callable, Generic, Optional, TypeVar
+
+from typing_extensions import Protocol
+
+import timemachine.lib.potentials as gpu
+import timemachine.potentials.chiral_restraints as ref_chiral
+from timemachine.potentials import bonded as ref_bonded
+from timemachine.potentials import nonbonded as ref_nonbonded
+
+Array = Any
+Conf = Array
+Params = Array
+Box = Array
+Lambda = float
+PotentialFxn = Callable[[Conf, Params, Box, Lambda], float]
+Precision = Any
```

Comment:
(Second tangent / nit: Is `Precision` type used here? I don't see a reference in this file, I didn't think Jax allowed setting precision more granularly than picking a single precision mode at runtime, and I thought precision would only be referenced later, `generic_potential.to_gpu().unbound_impl(np.float32)` or similar.)

Reply:
> Second tangent / nit: Is Precision type used here?

Good catch, this was left over from an earlier version where `to_gpu` accepted a `precision` argument and instantiated a `Potential` (I ended up opting for returning a `CustomOpWrapper` since that seemed simpler and more flexible for now). I removed the alias in 570f413002bbf7d3f546c3dd84447e414998accc.

> Is there a Pythonic way to define a type from explicitly enumerated values

I briefly considered introducing an [enum](https://docs.python.org/3/library/enum.html) for this, but it seemed more intrusive than made sense here.